File: summary/go/21_local_search.go
package main

import "fmt"

func localSearch(arr []int) []int {
	n := len(arr)
	bestSolution := make([]int, n)
	copy(bestSolution, arr)
	bestCost := calculateCost(bestSolution)

	for {
		improved := false
		currentSolution := make([]int, n)
		copy(currentSolution, bestSolution)

		for i := 0; i < n; i++ {
			for j := i + 1; j < n; j++ {
				currentSolution[i], currentSolution[j] = currentSolution[j], currentSolution[i]
				currentCost := calculateCost(currentSolution)
				if currentCost < bestCost {
					copy(bestSolution, currentSolution)
					bestCost = currentCost
					improved = true
				}
				currentSolution[i], currentSolution[j] = currentSolution[j], currentSolution[i]
			}
		}

		if !improved {
			break
		}
	}

	return bestSolution
}

func calculateCost(arr []int) int {
	cost := 0
	for _, value := range arr {
		cost += value * value
	}
	return cost
}

// Example usage:
func main() {
	input := []int{10, 20, 30, 40, 50}
	optimalArrangement := localSearch(input)
	fmt.Println("Optimized Array:", optimalArrangement)
}

File: summary/python/21_local_search.py
def local_search_algorithm(arr):
    n = len(arr)
    best_solution = arr.copy()
    best_cost = calculate_cost(best_solution)

    while True:
        improved = False
        for i in range(n):
            for j in range(i + 1, n):
                current_solution = best_solution.copy()
                current_solution[i], current_solution[j] = current_solution[j], current_solution[i]
                current_cost = calculate_cost(current_solution)
                if current_cost < best_cost:
                    best_solution = current_solution
                    best_cost = current_cost
                    improved = True
        
        if not improved:
            break

    return best_solution

def calculate_cost(arr):
    return sum(num ** 2 for num in arr)

# Example usage:
input_array = list(map(int, input().split()))
best_array = local_search_algorithm(input_array)
print("Optimized Array:", best_array)
